tests/azure_iot_provisioning_client_ut.c

```diff
@@ -0,0 +1,704 @@
+{
+    AzureIoTProvisioningClient_t xTestProvisioningClient;
+    AzureIoTMQTTPublishInfo_t xPublishInfo;
+    AzureIoTProvisioningClientResult_t xResult;
+
+    ( void ) ppvState;
+
+    prvSetupTestProvisioningClient( &xTestProvisioningClient );
+
+    /* Connect */
+    will_return( prvHmacFunction, 0 );
+    will_return( AzureIoTMQTT_Connect, eAzureIoTMQTTSuccess );
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 ),
+                      eAzureIoTProvisioningPending );
+
+    /* Subscribe */
+    will_return( AzureIoTMQTT_Subscribe, eAzureIoTMQTTSuccess );
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 ),
+                      eAzureIoTProvisioningPending );
+
+    /* AckSubscribe */
+    xPacketInfo.ucType = azureiotmqttPACKET_TYPE_SUBACK;
+    xDeserializedInfo.usPacketIdentifier = usTestPacketId;
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 ),
+                      eAzureIoTProvisioningPending );
+
+    /* Publish Registration Request */
+    will_return( AzureIoTMQTT_Publish, eAzureIoTMQTTSuccess );
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
+    xPacketInfo.ucType = 0;
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 ),
+                      eAzureIoTProvisioningPending );
+
+    /* Registration response */
+    prvGenerateGoodResponse( &xPublishInfo, 1 );
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 ),
+                      eAzureIoTProvisioningPending );
+
+    /* Read response */
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 ),
+                      eAzureIoTProvisioningPending );
+
+    /* Wait for timeout */
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
+    ullUnixTime += 2;
+    assert_int_equal( AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 ),
+                      eAzureIoTProvisioningPending );
+
+    /* Publish Registration Query */
+    will_return( AzureIoTMQTT_Publish, eAzureIoTMQTTSendFailed );
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
+    xPacketInfo.ucType = 0;
+    xResult = AzureIoTProvisioningClient_Register( &xTestProvisioningClient, 0 );
+    assert_int_not_equal( xResult, eAzureIoTProvisioningPending );
+    assert_int_not_equal( xResult, eAzureIoTProvisioningSuccess );
+
+    AzureIoTProvisioningClient_Deinit( &xTestProvisioningClient );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTProvisioningClient_Register_Success( void ** ppvState )
```

Comment:
OK if on later PR, but could we add a test here where we the Register passed in a long timeout.  I'm interested in getting some test over `prvProvClientRunWorkflow` path where it kicks into its do{}while loop >1 time and with permutations like the `(ulTimeoutMilliseconds% azureiotprovisioningPROCESS_LOOP_TIMEOUT_MS) = 0` and where say `(ulTimeoutMilliseconds% azureiotprovisioningPROCESS_LOOP_TIMEOUT_MS) != 0`. 

This would lend itself to helper functions for these various success cases.

Reply:
Yup

---
In reply to: [607973225](https://github.com/Azure/azure-iot-middleware-freertos/pull/62#discussion_r607973225) [](ancestors = 607973225)